### ISSUE TITLE
Write up the Secure Stop issue for ddorwin

### ISSUE THREAD

**Issue by slightlyoff** [w3ctag/spec-reviews#73]
@slightlyoff and @travisleithead to draft feedback regarding architecture impact/overlap of Secure Stop vs. `<a ping>` and `sendBeacon()`

**Comment by mwatson2** [comment]
Obviously, I wasn't at your meeting and I'm not sure from the notes what is to be written up. But I just wanted to make sure it was clear that the "write-on-close" approach taken, for example, by Firefox for this issue involves writing to persistent store (e.g. disk) at the time the browser tab is closed. Unlike <a ping> and sendBeacon() there is no network interaction required and no interaction with the page itself.

**Comment by mwatson2** [comment]
I heard the requirement described as needing to "write to disk after shutdown" and so I thought I should offer a small clarification.

Executing anything "after shutdown" is non-sensical, since by definition shutdown is complete at the point that execution stops. So nothing can be executed after that.

The particular implementation approach on which concerns have been raised is one in which the User Agent must write to disk during the process of unloading a document (i.e. as described at [1]) assuming that process takes place when the User Agent application is closed, as well as on navigation or tab or window close.

[1] http://dev.w3.org/html5/spec-LC/history.html#unloading-documents

**Comment by ddorwin** [comment]
The term I settled on was “write-on-application-close”.

More specifically, the requirement is that the web platform ensure that a write occurs during or sometime after the `MediaKeys` object is destroyed. That may very well occur during the algorithm to [unload a document](http://dev.w3.org/html5/spec-LC/history.html#unload-a-document), though it isn’t immediately clear to me where. I also wonder where objects that are not in the document structure (e.g. `MediaKeys` objects) are destroyed, and especially whether or how garbage collection affects this.

As far as spec language, I think that would need to be a blocking operation since - similar to your comments - the spec cannot specify what to do after the application has been destroyed.

**Comment by mwatson2** [comment]
Maybe splitting hairs, but rather than "after the `MediaKeys` object is destroyed" it should be "after all keys **known** to a `MediaKeySession` have been destroyed". CDMs could (and probably will) remove the keys before or at the same time as the `MediaKeySession` object and it's parent `MediaKeys` object are destroyed.

Since this is one implementation of a CDM-specific function, we won't need to describe it in the specifications, though discussing hypothetically how it *would* be described in our specifications may still be informative.

**Comment by ddorwin** [comment]
Yes, destruction of the `MediaKeySession` is a better way to represent it from the specification point of view. (I was thinking a little too much about implementation.)

If the observable behavior you want is for the last time to be written, then you do need to describe that behavior in the specification.

**Comment by ddorwin** [comment]
I just noticed that [1] points to a very old version of the spec. It should be http://www.w3.org/TR/html5/browsers.html#unloading-documents. I've also updated the link in my comment above. @mwatson2, you might want to do the same.

**Comment by travisleithead** [comment]
Hi folks, thanks for waiting. @slightlyoff and I have produced the following official TAG response:

**Comment by travisleithead** [comment]
## Architectural view on run-after-app-close behavior.

A question was posed to the TAG: some potentially synchronous operation are requested to be run "after application close", with clarifying statements about the durability, reliability, or timeframes that these operations are required to run within. 

The context for this question is described in [issue #85](https://github.com/w3c/encrypted-media/issues/85), a debate about the inclusion of a feature in EME that would specify shutdown behavior. In this response, we only seek to clarify  the architectural question of requiring steps to run after application close; we make no value judgement of the feature in question or implementation strategies vendors might choose.

The most basic app-close (e.g., "shutdown") activity that impacts the web platform is navigation, which causes documents to be created and destroyed. From the point of view of the document, its life-cycle starts after retrieving and parsing the resource's URL and ends when it is moved into the navigation history (when it is then "destroyed", but the exact timing of this is UA dependent). The management of documents and their life-cycles are provided by the browsing context. When a user closes a top-level browsing context (e.g., a browser tab) this results in the demise of both the current document (which should run unloading steps, as noted in the HTML specification, but might not) as well as its session history. Unlike documents, browsing contexts do not explicitly have life-cycle hooks.

When documents are destroyed, e.g. by navigation, JS objects are destructed. But this isn't the only time that can happen: in JavaScript the precise semantics of object destruction/collection are not made available to author code. Care is given in specs to make sure GC isn't made observable. Internal behavior may hold onto objects, but this is sort of "weird" from the layered-platform perspective. If something isn't in the retained object graph in JS, it's odd for it to "live" longer than references or the JS context.

Implementations may differ on the exact behavior of shutting down a browsing context, but most will start by unloading the document, following the steps described in the HTML specification that lead to the `beforeunload` event (so that the user can cancel the operation). Beyond that, the browsing context is essentially committed to shutting down and various optimizations are employed so that a minimal amount of subsequent code is run (for reasons of performance and battery life). Closing an entire browser (or any web platform host environment) is the sum of closing all of its top level browsing contexts and is an operation that, today, lives far outside the reach of specified behavior.

At least two other web platform features -- Beacons and `<a ping>` -- are intended to help developers avoid running their own application logic during shutdown. Neither feature requires browsers to run any particular code immediately during shutdown and both are "best effort" in major implementations today (despite what the specs might request). The [`HTMLAnchorElement`'s `ping` attribute](https://html.spec.whatwg.org/multipage/semantics.html#ping) sends a notice to the server regarding the anchor's navigation, and the [`sendBeacon` API](http://www.w3.org/TR/beacon/) allows for a [limited-size] payload, such as performance or analytic information, to be queued and relayed back to a server at a convenient time for the user agent.

These features address the desired reliable reporting in a best-effort way. They operate as late as possible in the life-cycle of the document, but use asynchronous techniques that are designed not to interfere or block navigation of a document nor shutdown of a browsing context. They also defer implementation details to UAs regarding the life-cycle of the browsing contexts. These are reasonable models to follow.

The major gap in providing opportunities to run arbitrary code (not declarative, canned behaviors) during browser shutdown is that there are no hooks to the life-cycle model of a browsing context spec'd anywhere in the web platform today. As cited in [issue 73](https://github.com/w3ctag/spec-reviews/issues/73#issuecomment-152110576), the unloading of a document is one possible place to trigger the operation, but it cannot be determined if this is the result of navigation (implying that the browsing context remains active) or closing of the browsing context itself. It may be possible to add hooks, however, as noted above, implementations are keen to avoid running any unnecessary code after the browsing context is committed to close. It's also unclear what context would execute the callbacks of such hooks in general. Must documents be left alive?

Given the desire to offload operations that delay navigation or shutdown into asynchronous and/or deferred actions, another related approach might be to design a "shutdown worker" which may be queued to start on navigation of a browsing context or launched at the user agent's earliest convenience (similar to the `sendBeacon` processing model). We leave such a design up to the interested reader but note that the upcoming [Background Sync](https://github.com/WICG/BackgroundSync/blob/master/explainer.md) work may provide a model.

Implementations are welcome to add triggers and hooks to run operations on shutdown of specific web platform environments, of course. It must be understood however, that many web platform environments may not offer the same opportunities; for example, implementations of the web platform that exist only embedded in silicon--they may be started when power is applied and stopped when power is removed. Other operating environments, especially resource-constrained environments like mobile phones may apply policies that constrain features like service workers or storage activity in order to save battery.

In conclusion, the TAG does not believe that a web-based feature should require executing steps in an environment that is already in the process of closing down as described above. In general, the TAG favors designs that promote asynchronous or deferred actions; in contrast, requiring run-at-close steps as requested would likely be synchronous in order to reliably work in such a scenario, and therefore not appropriate for the web platform.

**Comment by slightlyoff** [comment]
Hey all,

I've spoken with @travisleithead at this week's F2F TAG meeting and what we have come to understand by looking at this from many angles is that:

 - The bits of browsers that are defined by specs don't currently have any way of specing the behavior that Secure Stop implies, at least not the low-latency notification version that seems to be implied. But it's hard to know what's actually being requested by the proposed feature as it seems under-specified in this dimension. 
 - Without outlining a set of processes that live longer than documents, this feature looks strange and out of step with the platform.
 - While we'd obviously support explaining more of the platform, our best guess is that enumerating enough of the browser behavior to find a natural place for Secure Stop behavior will take several years of spec spelunking. The efforts that seems like they're in some way looking at this sort of outside-the-document behavior are very early (e.g. efforts to standardize Extensions).
 - Given that it doesn't make sense to include this feature if it can't be explained, our strong guidance is to move this to a V2 where it can receive the attention it really needs to be well-layered in the platform.

Thanks

**Comment by torgo** [comment]
Discussed at [Stockholm f2f](https://github.com/w3ctag/meetings/blob/gh-pages/2016/07-stockholm/agenda.md).

**Comment by travisleithead** [comment]
@slightlyoff and I discussed my last comment in https://github.com/w3c/encrypted-media/issues/85#issuecomment-232211115. We agree that the "secure stop" feature *can* be related (architecturally) to the existing [nebulously defined] a[ping] and sendBeacon(). Given we have no architectural concerns with a[ping] and sendBeacon (excepting lack of proper spec detail/available hooks for these out-of-band activities--which are not themselves architectural concerns), by extension we no longer have a concern with "secure stop" (persistent usage record). Closing this issue.

**Comment by xhwang-chromium** [comment]
Update on EME `persistent-usage-record` session:

- It's been brought back to the latest [EME editors draft](https://w3c.github.io/encrypted-media/#dom-mediakeysessiontype-persistent-usage-record) by the [Media Working Group](https://github.com/w3c/encrypted-media/pull/462).
- Since it's a clean reland, we'll reuse this thread as the TAG review for the feature to guide the implementation.

Please let me know if you have any comments or questions.

**Comment by dbaron** [comment]
I think it would be preferable to either file a new issue, or to at least fill out the [current issue template](https://raw.githubusercontent.com/w3ctag/design-reviews/master/.github/ISSUE_TEMPLATE/010-specification-review.md) in a comment in this issue.

**Comment by xhwang-chromium** [comment]
As requested, filling out the current issue template here:

**Title**: TAG review of EME "persistent-usage-record" session

**Summary**: Support a new MediaKeySessionType “persistent-usage-record session”, for which the license and key(s) are not persisted and for which a record of key usage is persisted when the keys available within the session are destroyed.

- Explainer¹ (minimally containing user needs and example code): https://github.com/xhwang-chromium/encrypted-media/blob/pur-session/persistent-usage-record-session-explainer.md
- Specification URL:
  * As in latest EME editors draft: https://w3c.github.io/encrypted-media/    
  * Diff against the current EME spec: https://github.com/w3c/encrypted-media/pull/462
- Tests: https://wpt.fyi/results/encrypted-media
- Security and Privacy self-review²: https://github.com/xhwang-chromium/encrypted-media/blob/pur-session/persistent-usage-record-session-tag-security-and-privacy.md
- GitHub repo: https://github.com/w3c/encrypted-media/
- Primary contacts (and their relationship to the specification):
  * Joey Parrish, @joeyparrish, [Google]
  * Greg Freedman, @gregwfreedman, [Netflix]
  * Xiaohan Wang, @xhwang-chromium, [Google]
  * Mark Watson, @mwatson2, [Netflix]
- Organization(s)/project(s) driving the specification: Google, Netflix
- Key pieces of existing multi-stakeholder review or discussion of this specification:
  * Previous TAG discussion: https://github.com/w3ctag/design-reviews/issues/73 
  * Previous W3C meeting discussion: https://www.w3.org/2019/07/02-me-minutes.html#item07
- External status/issue trackers for this specification (publicly visible, e.g. Chrome Status):
  * https://chromestatus.com/feature/5638708017496064

**Further details**:

  - [x] I have reviewed the TAG's [API Design Principles](https://w3ctag.github.io/design-principles/)
  - The group where the work on this specification is currently being done: Media Working Group
  - Major unresolved issues with or opposition to this specification: None

We'd prefer the TAG provide feedback as (please delete all but the desired option):

  💬 leave review feedback as a **comment in this issue** and @-notify @xhwang-chromium

**Comment by mwatson2** [comment]
Hi TAG, all,

I'm sure the WG is always happy to have any and all technical feedback, but
the feature has not changed at all since the last TAG review and (unless I
missed it) the WG has not asked for a TAG review, so I am not sure if one
is necessary ?

...Mark

**Comment by xhwang-chromium** [comment]
My understand is that my [previous comment](https://github.com/w3ctag/design-reviews/issues/73#issuecomment-642330564) still stands that we should just reuse this thread as the TAG review for the feature which already concluded (because as Mark pointed, the feature didn't change at all).

So there's no need for a new TAG review. I filled the issue template as requested only so that we have all the information in one place, not to start a new TAG review. Please let me know if my understanding is wrong, and if so, why a new TAG review is needed.

**Comment by dbaron** [comment]
I guess I'm confused as to what you were requesting in https://github.com/w3ctag/design-reviews/issues/73#issuecomment-642330564 .  That sounded to me that you were asking for a new review.

**Comment by xhwang-chromium** [comment]
Sorry for the confusion. It's just a FYI to use this thread as the TAG review for the feature, so we don't need a new TAG review, because the feature didn't change. Since we already have a conclusion on this this issue, we don't need any new work here. I hope this is clear now.

**Comment by dbaron** [comment]
So I think you're asking about the requirement in the [Blink process](https://www.chromium.org/blink/launching-features) that you get TAG review.  If that's the case -- then I'd note that the TAG is an elected body within the W3C and not part of the Blink or Chromium projects.  So if you're trying to get permission from the Blink project to skip requesting a TAG review on the basis of earlier discussion that happened in this issue, then I think you should be asking Blink folks for that, not the TAG.

**Comment by xhwang-chromium** [comment]
Thanks for the info. Indeed I am going through the Blink process. I'll see how that goes.
